### PR TITLE
Feature/35204 partial statuses

### PR DESCRIPTION
**Pull request by Bohdan-Medv**

### CODE REVIEW

File: src/Service/OrderHelper.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
+<?php declare(strict_types=1);
```
Comment (thomaswerth23) [inline src/Service/OrderHelper.php:1]: interesting that you declare strict, but did not do it on other files :P, also use a new line

File: src/Service/OrderHelper.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                    ),
+                ]
+            )
+        );
+
+        /** @var OrderEntity|null $order */
+        $order = $orderRepository->search($criteria, $context)->getEntities()->first();
+
+        if ($order === null) {
+            throw new InvalidTransactionException('');
```
Comment (thomaswerth23) [inline src/Service/OrderHelper.php:61]: Why no message?

File: src/Service/Payment.php [hunk 3]
```diff
@@ -443,4 +472,28 @@ public static function getAllowed(array $customFields): array
...
+     * @param string $goalStatus
+     * @return bool
+     */
+    public static function operationImossible(string $currentStatus, string $goalStatus): bool
+    {
+        if ($currentStatus === $goalStatus) {
+            return true;
+        }
+
+        if (!array_key_exists($currentStatus, Payment::POSSIBLE_STATUSES)) {
```
Comment (thomaswerth23) [inline src/Service/Payment.php:488]: is statuses really a word?
I think the plural for status is status or just use states :)

File: src/Service/PaymentHandler.php [hunk 4]
```diff
@@ -268,13 +275,13 @@ public function cancelPayment(string $hostedCheckoutId, int $amount, array $item
         $newStatus = $status;
         $amounts = [];
         $log = [];
+        $isFinal = $amount == $customFields[Form::CUSTOM_FIELD_WORLDLINE_PAYMENT_TRANSACTION_CAPTURE_AMOUNT];
```
Comment (thomaswerth23) [inline src/Service/PaymentHandler.php:278]: isFinal should be a bool, but it is probably an int, so the naming is strange here

File: src/Service/PaymentHandler.php [hunk 5]
```diff
@@ -640,160 +652,102 @@ private function updateDatabase(array $customFields): void
...
-                    0,
-                    ['status' => $statusCode, 'hostedCheckoutId' => $hostedCheckoutId]
-                );
-                $this->transactionStateHandler->process($orderTransactionId, $this->context);
-                break;
-            }
             case in_array($statusCode, Payment::STATUS_CAPTURED):
             {
-                if ($orderTransactionState === OrderTransactionStates::STATE_PAID) {
+                if (Payment::operationImossible($orderTransactionState, OrderTransactionStates::STATE_PAID)) {
```
Comment (thomaswerth23) [inline src/Service/PaymentHandler.php]: typo Impossible instead of Imossible

File: src/Service/OrderHelper.php (hunk 1 above)
Comment (Bohdan-Medv) [reply src/Service/OrderHelper.php:1]: Here I copy SW6 service structure, will add strict to other files.

No new line in any of SW6 files and example plugins, looks like internal formating.

File: src/Service/OrderHelper.php (hunk 2 above)
Comment (Bohdan-Medv) [reply src/Service/OrderHelper.php:61]: Standard message is 'The transaction with id {{ transactionId }} is invalid or could not be found.'
At this point we have only external transaction ID, providing it here no make sense.

We will face this exception on Worldline malfunction or wrong return URL config, so external transaction ID will not help.
But I'll add it to log in refactoring merge request.

File: src/Service/Payment.php (hunk 3 above)
Comment (Bohdan-Medv) [reply src/Service/Payment.php:488]: the plural of status can simply be status (with no change) or statuses (with an “es”).

That means either of the following sentences would be acceptable:

    We polled a thousand Americans and had them respond with their economic status.

    Each of the countries in question filed reports on the statuses of their internet infrastructure.

**Review by thomaswerth23 — APPROVED** [review]

**Pull request merged**